File: Data Structures & Algorithms/maximum-subarray/submission-3.java
class Solution {
    public int maxSubArray(int[] nums) {
        int sum = 0, l = 0, r = 0, n = nums.length;
        int ans = nums[0];

        while(r<n){
            sum+=nums[r];
            ans = Math.max(ans, sum);
            if(sum <0) {
                sum = 0;
                l = r;
            } 
            r++;
        }
        return ans;
    }
}

File: Data Structures & Algorithms/maximum-product-subarray/submission-3.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        int ans = nums[0];
        int mini = 1, maxi = 1;

        for(auto x: nums) {
            int curr = maxi * x;
            maxi = max({x, maxi*x, mini*x });
            mini = min({x, curr, mini*x });
            ans = max(ans, maxi);
        }
        return ans;
    }
};
